scripts/backport.py

```diff
@@ -0,0 +1,371 @@
+            "and BACKPORT_TARGET_REMOTE (e.g. `origin`).",
+            file=sys.stderr,
+        )
+        sys.exit(1)
+
+print(
+    f"Will look at '{source_repo_name}' (git remote '{source_remote}') for bug fixes, "
+    f"and create the backport PRs in '{target_repo_name}' (git remote '{target_remote}')."
+)
+
+source_repo = github.get_repo(source_repo_name)
+target_repo = github.get_repo(target_repo_name)
+
+# Set git name and email corresponding to the token user.
+token_user = github.get_user()
+os.environ["GIT_COMMITTER_NAME"] = token_user.name
+os.environ[
+    "GIT_COMMITTER_EMAIL"
+] = f"{token_user.id}+{token_user.login}@users.noreply.github.com"
+print(
+    f"Will commit as {os.environ['GIT_COMMITTER_NAME']} <{os.environ['GIT_COMMITTER_EMAIL']}>"
+)
+
+# Fetch the sources
+subprocess.run(f"git fetch {source_remote}", shell=True, check=True)
+subprocess.run(f"git fetch {target_remote}", shell=True, check=True)
+
+# Find out what is the branch corresponding to the previous version compared to
+# main. We will backport to that branch.
+version_config = {
+    x[0].strip(): x[1].strip()
+    for x in [
+        line.split("=")
+        for line in subprocess.check_output(
+            f"git show {source_remote}/main:version.config",
+            shell=True,
+            text=True,
+        ).splitlines()
+        if line
+    ]
+}
+
+previous_version = version_config["update_from_version"]
+previous_version_parts = previous_version.split(".")
+previous_version_parts[-1] = "x"
+backport_target = ".".join(previous_version_parts)
+
+print(f"Will backport to {backport_target}")
+
+# Find out which commits are unique to main and target branch. Also build sets of
+# the titles of these commits. We will compare the titles to check whether a
+# commit was backported.
+main_commits = [
+    line.split("\t")
+    for line in subprocess.check_output(
+        f'git log -1000 --pretty="format:%h\t%s" {source_remote}/{backport_target}..{source_remote}/main',
+        text=True,
+        shell=True,
+    ).splitlines()
+    if line
+]
+main_titles = {x[1] for x in main_commits}
+
+branch_commits = [
+    line.split("\t")
+    for line in subprocess.check_output(
+        f'git log -1000 --pretty="format:%h\t%s" {source_remote}/main..{source_remote}/{backport_target}',
+        text=True,
+        shell=True,
+    ).splitlines()
+    if line
+]
+branch_titles = {x[1] for x in branch_commits}
+
+# We will do backports per-PR, because one PR, though not often, might contain
+# many commits. So as the first step, go through the commits unique to main, find
+# out which of them have to be backported, and remember the corresponding PRs.
+# We also have to remember which commits to backport. The list from PR itself is
+# not what we need, these are the original commits from the PR branch, and we
+# need the resulting commits in master. Store them as dict(pr number -> PRINfo).
```

Comment:
Hmm.... quite often PRs contain multiple commit messages because they are fixing separate issues but are added to a single PR for convenience. In that case, it might make sense to cherry-pick them separately and create a PR for each.

Reply:
I think we can't really handle that, because we can't understand which commit fixes which issue. One might need to be backported, and the other not. I can disallow PRs with multiple commits.

Reply:
> I can skip PRs with multiple commits.

Did that.

Reply:
> I think we can't really handle that, because we can't understand which commit fixes which issue.

Why not? Each commit would contain a "fixed" label in this case, so it should be straightforward to cherry-pick these and create separate PRs?

> One might need to be backported, and the other not. I can skip PRs with multiple commits.

Not sure why one should be backported and the other one not. Both are committed in a PR, so both should be part of the release.

Reply:
> Not sure why one should be backported and the other one not. Both are committed in a PR, so both should be part of the release.

If both should be backported together, separate PRs seem to contradict it? 

In the case you described above, we have two commits fixing two different issues in one PR. The current rule for backporting is "if the fixed issue is labeled 'bug' and not labeled 'no-backport'". These labels can differ for these issues, so the decision to backport will also differ.

I think it is more convenient to make backport decision on a per-PR basis. Commits often don't have "fixes" references (and I think it's good, because they tend to spam the issues endlessly at each rebase). We will also need more code to support both per-PR and per-commit basis, and pretty tricky logic to combine these approaches, because they might easily give different results.

Reply:
> In the case you described above, we have two commits fixing two different issues in one PR. The current rule for backporting is "if the fixed issue is labeled 'bug' and not labeled 'no-backport'". These labels can differ for these issues, so the decision to backport will also differ.

But then they should not be merged as a single PR. They should actually be separate PRs.

To be able to backport on a per-PR basis, it is important that if a PR consists of multiple commits, it is because they are **impossible** to submit as separate PRs. In that case, and only in that case, you can safely assume that the entire PR should be backported.

Reply:
> Commits often don't have "fixes" references (and I think it's good, because they tend to spam the issues endlessly at each rebase).

All mine have. :)

Reply:
> I think it is more convenient to make backport decision on a per-PR basis. [...] We will also need more code to support both per-PR and per-commit basis, and pretty tricky logic to combine these approaches, because they might easily give different results.

I am leaning towards the per-PR basis for another reason than "more code to support both". It is in some situations necessary to have two commits for a single change. For example, if you rename a lot of files and then modify them, it is better to have one commit for the renaming and one for the following changes. In this case, the PR would be the glue for these two PRs.